src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -887,6 +895,62 @@ bool ClusterTaskManager::ReturnCpuResourcesToBlockedWorker(
 void ClusterTaskManager::ScheduleAndDispatchTasks() {
   SchedulePendingTasks();
   DispatchScheduledTasksToWorkers(worker_pool_, leased_workers_);
+  // TODO(swang): Spill from waiting queue first? Otherwise, we may end up
+  // spilling a task whose args are already local.
+  // TODO(swang): Invoke ScheduleAndDispatchTasks() when we run out of memory
+  // in the PullManager or periodically, to make sure that we spill waiting
+  // tasks that are blocked.
+  SpillWaitingTasks();
+}
+
+void ClusterTaskManager::SpillWaitingTasks() {
+  // Try to spill waiting tasks to a remote node, prioritizing those at the end
+  // of the queue. Waiting tasks are spilled if there are enough remote
+  // resources AND (we have no resources available locally OR their
+  // dependencies are not being fetched). We should not spill tasks whose
+  // dependencies are actively being fetched because some of their dependencies
+  // may already be local or in-flight to this node.
+  //
+  // NOTE(swang): We do not iterate by scheduling class here, so if we break
+  // due to lack of remote resources, it is possible that a waiting task that
+  // is earlier in the queue could have been scheduled to a remote node.
+  auto it = waiting_task_queue_.end();
+  while (it != waiting_task_queue_.begin()) {
+    it--;
+    const auto &task = std::get<0>(*it);
+    const auto &task_id = task.GetTaskSpecification().TaskId();
+    // Check whether this task's dependencies are not being pulled.  If this is
+    // true, then we should force the task onto a remote feasible node, even if
+    // we have enough resources available locally for placement.
+    bool force_spillback = !task_dependency_manager_.TaskDependenciesPending(task_id);
```

Comment:
```suggestion
    bool force_spillback = task_dependency_manager_.CannotExecuteLocallyYet(task_id);
```

Reply:
(avoid double negative)

Reply:
I don't think this name is accurate. The call just returns whether dependencies are being fetched, not whether the dependencies are local yet.

Reply:
(also avoid "pending" since arguably everything is "pending" in the task dep manager)

Reply:
How about "TaskDependenciesBlocked" then?

Reply:
:+1: blocked